Added functionality for PT order creation from UI in Payment step

abhishekgupta5 opened the pull request:
**Issue:** https://linear.app/nebulab-retainers/issue/ABU-7/create-a-template-controller-to-create-paytomorrow-order

- [x] Added controller that calls **OrderCreationService**
- [x] Added JS script to handle Pay on payment step and call^
- [ ] Make it a template controller
- [x] Added partial to show PT on the payment step with 'pay' button (__pay_tomorrow.html.erb_)

**TODO -**
1. The success/fail HTML after user Clicks **Pay** should be made beautiful. It's plain rn

abhishekgupta5 committed:
Added SolidusPayTomorrow::Api::CreateOrderController

1. The API controller is called from the Payment step while user
   is paying via PT.
   It returns success/failure json and the JS handles the redirect on
   frontend

2. BaseController inherits from ::Spree::Api::BaseController so that
   we can leverage Solidus' core auth methods

3. Generic error handling in for controllers inheriting from
   BaseController

abhishekgupta5 committed:
Added errorDetails key in error details

PT gives an error description in this attribute sometimes.
Eg - If while creating an order, last name is not provided,
     it results in such a response

 ```
   {
     "timestamp": "2022-10-06T11:44:28.559Z",
     "status": "EXPECTATION_FAILED",
     "path": "/application/ecommerce/orders",
     "error": "Expectation Failed",
     "errorDetails": "Seems like the json you sent is missing info or has the wrong format:
                     - property lastName must not be null, you sent: 'null'" }
```

This change is done so that we may provide a more verbose error
message

DanielePalombo commented on app/views/spree/checkout/payment/_pay_tomorrow.html.erb:
```diff
@@ -0,0 +1,5 @@
+<p> Payment method description: <%= payment_method.description %></p>
+
+<br>
+<%= link_to 'Pay', create_pt_order_path(order_number: @current_order.number, payment_method: ), method: 'post', class: 'button' %>
```
Why this is still using `pt` rather than `PayTomorrow` naming? 🤔

abhishekgupta5 replied:
Because the route is aliased to `create_pt_order` in `routes.rb`. Shall I make this `create_pay_tomorrow_order_path`?

DanielePalombo replied:
Yes, I think it's better to keep the fullname of the service here

abhishekgupta5 replied:
Fixed in dcf1962452ad25e8a7d4054997756061073d561b

DanielePalombo commented on app/controllers/solidus_pay_tomorrow/base_controller.rb:
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+module SolidusPayTomorrow
+  class BaseController < ::Spree::Api::BaseController
```
Why the BaseContoller inherits from the API::BaseController ? 🤔

abhishekgupta5 replied:
If I don't use `skip_before_action` hooks defined in `::Spree::Api::BaseController`, I get - 
`You must specify an API key` hence I have inherited from  `::Spree::Api::BaseController` after taking reference from [Bolt](https://github.com/solidusio/solidus_bolt/blob/master/app/controllers/solidus_bolt/base_controller.rb)

DanielePalombo replied:
Yes, but here the purpose is different. In Bolt we skipped the authentication because the call came from their services, and they can't be authorized on our side with the bearer authentication. Still, as you can see, there is a [verify_bolt_request](https://github.com/solidusio/solidus_bolt/blob/master/app/controllers/solidus_bolt/base_controller.rb#L13-L19) method responsible for verifying that the request is made by Bolt.
The API here is used by our FE, and we should use the guest token to authorize it.

abhishekgupta5 replied:
Added guest_token in https://github.com/nebulab/solidus_pay_tomorrow/pull/9/commits/7af9fc62fba5a622a479931f394c675b2183f422

DanielePalombo commented on app/controllers/solidus_pay_tomorrow/create_order_controller.rb:
```diff
@@ -0,0 +1,23 @@
+        render 'spree/checkout/payment/_checkout', status: :created
+      end
+    rescue StandardError => _e
+      render 'spree/checkout/payment/_checkout', status: :unprocessable_entity
```
Can we add the error message somewhere?

abhishekgupta5 replied:
Sure. Added in https://github.com/nebulab/solidus_pay_tomorrow/pull/9/commits/6dbea12df49f94581115e9fb4a14c0bc673215b0

DanielePalombo commented on config/routes.rb:
```diff
@@ -1,5 +1,5 @@
 
 Spree::Core::Engine.routes.draw do
-  # Add your extension routes here
+  post '/api/order/pay_tomorrow', to: '/solidus_pay_tomorrow/create_order#create', as: 'create_pt_order'
```
Can we use the resources routes convention here?

something like:

```
scope 'solidus_pay_tomorrow' do
  scope 'api' do
    resources 'orders', only: [:create]
  end
end
```

abhishekgupta5 replied:
Will have to decide on https://github.com/nebulab/solidus_pay_tomorrow/pull/9#discussion_r989087885 in order to change this. API controller shall have `api` scope, not FE controller.

abhishekgupta5 replied:
I doubt we can use **resources** method here. Actions defined under resources are being redirected to `Spree::...` controllers.

For eg - Above resource points to `spree/orders_orders#create`

-----
Also, **resource#create** denotes the creation of the resource record in the rails app, isn't it?
Here the purpose of the this controller is to create PayTomorrow order externally, we don't actually create the `Order` object, hence it may be confusing.

DanielePalombo commented on app/controllers/solidus_pay_tomorrow/create_order_controller.rb:
```diff
@@ -0,0 +1,23 @@
+module SolidusPayTomorrow
+  class CreateOrderController < BaseController
+    def create
+      order = Spree::Order.find_by(number: permitted_params[:order_number])
```
This doesn't look right to me, because it will allow any user with the order_number to start the checkout flow on PayTOmorrow.
It's not so clear if this controller should be a FrontendController or an APIController.
For frontend controller I think you can check the current_order, and the order will be retrieved by the cookies. In the order case(if this should be an API controller) you have to pass the guest_token as x-spree-order-token in the header.

abhishekgupta5 replied:
~**Q:** Since this is expected to be done from the Solidus app only, I think this should be a FE controller, not API controller. Wdyt?~

Answer is No, because of the comment below

abhishekgupta5 replied:
We decided API controller will be better as we'd want to avoid FE controller because of _solidus_starter_frontend_ transition

abhishekgupta5 replied:
Added guest_token in https://github.com/nebulab/solidus_pay_tomorrow/pull/9/commits/7af9fc62fba5a622a479931f394c675b2183f422

DanielePalombo commented on app/controllers/solidus_pay_tomorrow/create_order_controller.rb:
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+module SolidusPayTomorrow
+  class CreateOrderController < BaseController
```
This Controller should be under `Api` module

DanielePalombo replied:
Nevermind, I read the routes and I thought it should be under the `API`

DanielePalombo commented on config/routes.rb:
```diff
@@ -1,5 +1,5 @@
 
 Spree::Core::Engine.routes.draw do
-  # Add your extension routes here
+  post '/api/order/pay_tomorrow', to: '/solidus_pay_tomorrow/create_order#create', as: 'create_pay_tomorrow_order'
```
~I think the right path should be:~

~`/api/pay_tomorrow/order`, and the name should be `create_api_pay_tomorrow_order`~

NVM

DanielePalombo replied:
I think the API namespace here is a typo

DanielePalombo replied:
This is not clear, because the Controller inherits from Spree::API I'm resolving this conversation and I'll try to provide more details on the Requested change message.

DanielePalombo commented on app/controllers/solidus_pay_tomorrow/create_order_controller.rb:
```diff
@@ -0,0 +1,24 @@
+module SolidusPayTomorrow
+  class CreateOrderController < BaseController
+    def create
+      order = Spree::Order.find_by(number: permitted_params[:order_number])
```
You added the order_token, this is no more needed because the order retrieved by the guest_token should be available on @order instance variable.

abhishekgupta5 replied:
Used **@order_token** directly now

DanielePalombo commented on app/controllers/solidus_pay_tomorrow/create_order_controller.rb:
```diff
@@ -0,0 +1,24 @@
+      @result = SolidusPayTomorrow::Client::CreateOrderService.call(order: order, payment_method: payment_method)
+      @successfully_created = @result.key?('url') && @result.key?('token')
+      if @successfully_created
+        render 'spree/checkout/payment/_checkout', status: :created
```
This controller inherits from SolidusPayTomorrow::BaseController that inherits from Spree::Api::BaseController, this means we should render a JSON rather an HTML code.

abhishekgupta5 replied:
Changed in https://github.com/nebulab/solidus_pay_tomorrow/pull/9/commits/bd16050623ac9409f748fc49c53b2056f9636922

DanielePalombo commented on app/controllers/solidus_pay_tomorrow/create_order_controller.rb:
```diff
@@ -0,0 +1,24 @@
+      end
+    rescue StandardError => e
+      @error_message = e.message
+      render 'spree/checkout/payment/_checkout', status: :unprocessable_entity
```
It should be a JSON

abhishekgupta5 replied:
Changed in https://github.com/nebulab/solidus_pay_tomorrow/pull/9/commits/bd16050623ac9409f748fc49c53b2056f9636922

DanielePalombo commented on app/controllers/solidus_pay_tomorrow/create_order_controller.rb:
```diff
@@ -0,0 +1,24 @@
+      if @successfully_created
+        render 'spree/checkout/payment/_checkout', status: :created
+      end
+    rescue StandardError => e
```
Considering it surrounds the only available action, we can probably use `rescue_from` on the controller.

abhishekgupta5 replied:
Will do

abhishekgupta5 replied:
Done in https://github.com/nebulab/solidus_pay_tomorrow/pull/9/commits/bd16050623ac9409f748fc49c53b2056f9636922

DanielePalombo reviewed (changes requested):
I think there was some miscommunication during the last call.

As far as I remember we decided to add a JS code to inquiry the API responsible for creating the order in PayTomorrow and then to redirect the user to the given URL.

This means I would expect the following:

1) API controller (under SolidusPayTomorrow::Api namespace) that call CreateOrderService and return back the URL in JSON
2) A js code that handles the Click of the user on the `Pay` button, calls the above Controller and redirects the user to the given URL

I think you can get some inspiration from reading https://github.com/solidusio-contrib/solidus_klarna_payments/blob/master/app/controllers/solidus_klarna_payments/api/callbacks_controller.rb

DanielePalombo commented on spec/requests/solidus_pay_tomorrow/create_order_controller_spec.rb:
```diff
@@ -0,0 +1,31 @@
+        )
+    end
+
+    context 'when order is created successfully' do
+      it 'returns created status code' do
+        call
+        expect(response).to have_http_status(:created)
+      end
+    end
+  end
```
You added the rescue, so you should add a test for it.

abhishekgupta5 replied:
Will add

abhishekgupta5 replied:
Added in https://github.com/nebulab/solidus_pay_tomorrow/pull/9/commits/bd16050623ac9409f748fc49c53b2056f9636922

abhishekgupta5 commented:
> I think there was some miscommunication during the last call.
> 
> As far as I remember we decided to add a JS code to inquiry the API responsible for creating the order in PayTomorrow and then to redirect the user to the given URL.
> 
> This means I would expect the following:
> 
>     1. API controller (under SolidusPayTomorrow::Api namespace) that call CreateOrderService and return back the URL in JSON
> 
>     2. A js code that handles the Click of the user on the `Pay` button, calls the above Controller and redirects the user to the given URL
> 
> 
> I think you can get some inspiration from reading https://github.com/solidusio-contrib/solidus_klarna_payments/blob/master/app/controllers/solidus_klarna_payments/api/callbacks_controller.rb

------------------

@DanielePalombo  what I understood from our call, I tried to confirm [here on Linear](https://linear.app/nebulab-retainers/issue/ABU-7#comment-573a2b9a) . I am re-posting the same

_We decided to use JS request to create the PT order instead of using a rails controller.
Now that I think about this - what will be the use of [OrderCreationService](https://github.com/nebulab/solidus_pay_tomorrow/blob/master/app/services/solidus_pay_tomorrow/client/create_order_service.rb) if we'll create order from JS?
Also, we'll have to ensure that the access_token is valid here in JS request as well. Basically, whatever [OrderCreationService](https://github.com/nebulab/solidus_pay_tomorrow/blob/master/app/services/solidus_pay_tomorrow/client/create_order_service.rb) and [BaseService](https://github.com/nebulab/solidus_pay_tomorrow/blob/master/app/services/solidus_pay_tomorrow/client/base_service.rb) handle for us, we'll have to replicate in JS_

------------------

Now ->

From the [quoted/requested comment](https://github.com/nebulab/solidus_pay_tomorrow/pull/9#pullrequestreview-1138149399), I feel what you are trying to say is - 

1. Put `CreateOrderController` under API namespace hence it should be `SolidusPayTomorrow::Api::CreateOrderController` and it should return the URL and token in json
2. When user clicks Pay, instead of calling this controller directly, we need a JS function that calls this controller, get the above^ json, redirects the user to the url


Is my understanding correct? If so, what else the JS is supposed to do apart from this^

DanielePalombo commented:
The user should be redirected by pay tomorrow after the application is created

abhishekgupta5 commented:
> The user should be redirected by pay tomorrow after the application is created

Right. To the returnURL.
And we'd have a new controller to handle that?

DanielePalombo commented:
> And we'd have a new controller to handle that?

Yes, but this won't be part of this PR.

abhishekgupta5 committed:
Added HTML/JS that will call CreateOrderController

1. When user will click on the Pay button on the payment step,
   the JS will do the post request to CreateOrderController,
   fetch the url and token and redirect the user to
   Paytomorrow's URL
2. We disable the Pay button when clicked once so that,
   we can avoid duplicate requests
3. We show a very basic success/failure message on Frontend
   based on CreateOrderController's response

abhishekgupta5 committed:
Renamed partial_name

From 'pt' to 'pay_tomorrow' for clarity

abhishekgupta5 commented:
> I think there was some miscommunication during the last call.
> 
> As far as I remember we decided to add a JS code to inquiry the API responsible for creating the order in PayTomorrow and then to redirect the user to the given URL.
> 
> This means I would expect the following:
> 
>     1. API controller (under SolidusPayTomorrow::Api namespace) that call CreateOrderService and return back the URL in JSON
> 
>     2. A js code that handles the Click of the user on the `Pay` button, calls the above Controller and redirects the user to the given URL
> 
> 
> I think you can get some inspiration from reading https://github.com/solidusio-contrib/solidus_klarna_payments/blob/master/app/controllers/solidus_klarna_payments/api/callbacks_controller.rb

Fixed all of this in https://github.com/nebulab/solidus_pay_tomorrow/pull/9/commits/588020bf4ae3b46ecc697ead0d662b656f41ddae

DanielePalombo reviewed (approved):
Tested with @abhishekgupta5 and looks good. 

There are some unexpected issue but they look related to solidus.

DanielePalombo committed:
Move js to the right file

DanielePalombo committed:
Adjust the sandbox bin to the latest version

DanielePalombo committed:
Fix preferences issue

This commit changes payment_method.preferences[:signature] to
payment_method.preferred_signature in order to fix the error

DanielePalombo committed:
Add SolidusPayTomorrow::PaymentMethod seed

This commit adds a seed to create a PayTomorrow Payment if the required
keys are present on the environment.

DanielePalombo commented:
@abhishekgupta5 as discussed IRL, I added some commits to fix some problems with the sandbox and make the dev process more smooth. If the PR looks good to you, please merge it.
